# Data Diggers Team Assignment Analytics Applications

<a href="https://colab.research.google.com/github/asimxa/Data-Diggers-Project-Repo/blob/main/Team_Assignment_24_25.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [111]:
# packages
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

In [112]:
# read csv files
try:
    charging_sessions_df = pd.read_csv('charging_sessions.csv')
    print("Successfully imported charging_sessions.csv")
except FileNotFoundError:
    print("Error: charging_sessions.csv not found.")

try:
    weather_burbank_airport_df = pd.read_csv('weather_burbank_airport.csv')
    print("Successfully imported weather_burbank_airport.csv")
except FileNotFoundError:
    print("Error: weather_burbank_airport.csv not found.")

Successfully imported charging_sessions.csv
Successfully imported weather_burbank_airport.csv


In [113]:
charging_sessions_df

Unnamed: 0                        id             connectionTime  \
0               0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1               1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2               2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3               3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4               4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...           ...                       ...                        ...   
66445       10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446       10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447       10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448       10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449       10087  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN  

[66450 rows x 13 columns]

## Data Collection and Preparation

In the upcoming section, we want to observe how many erroneous datasets we have and how we should handle them.

### Format
We first observe that the .csv file's first column is an unnamed column with integers. Since there is no description in the .pdf file for the team assignment, we will drop these incoherent values and get rid of the very first column.

In [114]:
formatted_df = charging_sessions_df.copy().iloc[:, 1:]
formatted_df

id             connectionTime  \
0      5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1      5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2      5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3      5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4      5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...                         ...                        ...   
66445  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN  

[66450 rows x 12 columns]

Now we take a look at the datatypes in our dataframe:

In [115]:
print(formatted_df.dtypes)

id                   object
connectionTime       object
disconnectTime       object
doneChargingTime     object
kWhDelivered        float64
sessionID            object
siteID                int64
spaceID              object
stationID            object
timezone             object
userID              float64
userInputs           object
dtype: object


For easier handling during our analysis, it might be benificial to convert the values of **connectionTime**, **disconnectTime** and **doneChargingtime** from type object to datetime.

In [121]:
formatted_df['connectionTime'] = pd.to_datetime(formatted_df['connectionTime'], utc=False, errors='coerce')
formatted_df['disconnectTime'] = pd.to_datetime(formatted_df['disconnectTime'], utc=False, errors='coerce')
formatted_df['doneChargingTime'] = pd.to_datetime(formatted_df['doneChargingTime'], utc=False, errors='coerce')
print(formatted_df.dtypes)

id                               object
connectionTime      datetime64[ns, UTC]
disconnectTime      datetime64[ns, UTC]
doneChargingTime    datetime64[ns, UTC]
kWhDelivered                    float64
sessionID                        object
siteID                            int64
spaceID                          object
stationID                        object
timezone                         object
userID                          float64
userInputs                       object
dtype: object


### Duplicates
First we want to find out whether any duplicates exist which we can check via the **id** field of each dataset.
If so, those duplicates shall be removed.

In [6]:
len(formatted_df[formatted_df['id'].duplicated()])

1413

As we can see, there are **1413** duplicated rows. We will thus remove them for further analysis.

In [7]:
formatted_df = formatted_df.drop_duplicates(subset=['id'])
formatted_df

id             connectionTime  \
0      5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1      5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2      5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3      5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4      5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...                         ...                        ...   
65032  5d2fbdd3f9af8b4d0dd0d54a  2019-07-01 21:49:12+00:00   
65033  5d2fbdd3f9af8b4d0dd0d54b  2019-07-01 21:58:45+00:00   
65034  5d2fbdd3f9af8b4d0dd0d54c  2019-07-01 22:02:21+00:00   
65035  5d2fbdd3f9af8b4d0dd0d54d  2019-07-01 22:23:44+00:00   
65036  5d2fbdd3f9af8b4d0dd0d54e  2019-07-01 22:36:08+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
65032  2019-07-02 01:37:18+00:00  2019-07-01 23:14:33+00:00         8.399   
65033  2019-07-02 00:39:48+00:00  2019-07-02 00:40:21+00:00        16.864   
65034  2019-07-02 00:58:50+00:00  2019-07-02 00:59:23+00:00        18.335   
65035  2019-07-02 02:03:36+00:00  2019-07-02 02:04:06+00:00        22.815   
65036  2019-07-02 01:10:52+00:00  2019-07-02 00:24:32+00:00        10.241   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
65032  1_1_179_798_2019-07-01 21:49:11.873404       1  AG-3F18  1-1-179-798   
65033  1_1_179_794_2019-07-01 21:58:44.571011       1  AG-3F20  1-1-179-794   
65034  1_1_191_807_2019-07-01 22:02:20.810735       1  AG-4F47  1-1-191-807   
65035  1_1_179_781_2019-07-01 22:23:32.496137       1  AG-3F31  1-1-179-781   
65036  1_1_193_819_2019-07-01 22:35:56.104381       1  AG-1F06  1-1-193-819   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
65032  America/Los_Angeles  1346.0   
65033  America/Los_Angeles   364.0   
65034  America/Los_Angeles  2050.0   
65035  America/Los_Angeles  1626.0   
65036  America/Los_Angeles  2299.0   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
65032  [{'WhPerMile': 400, 'kWhRequested': 32.0, 'mil...  
65033  [{'WhPerMile': 400, 'kWhRequested': 40.0, 'mil...  
65034  [{'WhPerMile': 333, 'kWhRequested': 29.97, 'mi...  
65035  [{'WhPerMile': 200, 'kWhRequested': 34.0, 'mil...  
65036  [{'WhPerMile': 333, 'kWhRequested': 16.65, 'mi...  

[65037 rows x 12 columns]

### Missing Data
We now want to check the data for missing values

In [24]:
formatted_df.isnull().sum()

id                      0
connectionTime          0
disconnectTime          0
doneChargingTime     4087
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID              17215
userInputs          17215
dtype: int64

As we can see, there are **4087** rows with a missing **doneChargingTime** value as well as **17215** rows with missing **userID** and **userInputs** value. The latter is expected as charging sessions may also be carried out by unregistered users. Thus, we only have to take care of the **doneChargingTime** values. We try to find out whether this is related to a specific other value.

In [41]:
missing_done_charging_time_df = formatted_df[formatted_df.doneChargingTime.isnull()]
space_id_value_counts = missing_done_charging_time_df['stationID'].value_counts()
space_id_value_counts

stationID
2-39-81-4550    1125
1-1-178-823       89
1-1-193-820       77
1-1-178-828       67
1-1-193-827       63
                ... 
2-39-129-563       2
2-39-130-564       1
2-39-125-559       1
2-39-126-560       1
2-39-82-384        1
Name: count, Length: 100, dtype: int64

The stationID **2-39-81-4550** seems to be faulty when it comes to this value. When we look at its records:

In [52]:
print("stationID 2-39-81-4550")
print(f"null values: {missing_done_charging_time_df[missing_done_charging_time_df.stationID == '2-39-81-4550']['doneChargingTime'].isnull().sum()}")
print(f"non null values: {missing_done_charging_time_df[missing_done_charging_time_df.stationID == '2-39-81-4550']['doneChargingTime'].notnull().sum()}")

stationID 2-39-81-4550
null values: 1125
non null values: 0


In [53]:
missing_done_charging_time_df[missing_done_charging_time_df.stationID == '2-39-81-4550']

id             connectionTime  \
46086  6004e49bf9af8b546d1d6e03  2021-01-01 22:29:34+00:00   
46089  6004e49bf9af8b546d1d6e06  2021-01-02 02:03:29+00:00   
46090  6004e49bf9af8b546d1d6e07  2021-01-02 02:53:51+00:00   
46091  6004e49bf9af8b546d1d6e08  2021-01-02 04:34:46+00:00   
46092  6006361cf9af8b569d876177  2021-01-02 15:07:37+00:00   
...                         ...                        ...   
51579  6000f01bf9af8b4e2987106a  2020-12-30 06:35:39+00:00   
51580  6002419cf9af8b50514f4c3c  2020-12-30 18:17:31+00:00   
51582  6002419cf9af8b50514f4c3e  2020-12-30 22:09:41+00:00   
51584  6003931bf9af8b5261c8c793  2020-12-31 16:17:43+00:00   
51586  6003931bf9af8b5261c8c795  2020-12-31 17:42:03+00:00   

                  disconnectTime doneChargingTime  kWhDelivered  \
46086  2021-01-01 23:20:56+00:00              NaN        35.503   
46089  2021-01-02 02:19:50+00:00              NaN        11.827   
46090  2021-01-02 03:23:48+00:00              NaN         7.646   
46091  2021-01-02 05:22:23+00:00              NaN        31.138   
46092  2021-01-02 15:55:46+00:00              NaN        12.768   
...                          ...              ...           ...   
51579  2020-12-30 07:30:14+00:00              NaN        30.430   
51580  2020-12-30 19:09:46+00:00              NaN        29.623   
51582  2020-12-30 22:33:48+00:00              NaN         6.331   
51584  2020-12-31 17:09:35+00:00              NaN        24.903   
51586  2020-12-31 18:29:00+00:00              NaN        20.526   

                                     sessionID  siteID   spaceID  \
46086  2_39_81_4550_2021-01-01 22:29:30.587407       2  11900388   
46089  2_39_81_4550_2021-01-02 02:03:26.186580       2  11900388   
46090  2_39_81_4550_2021-01-02 02:53:45.713992       2  11900388   
46091  2_39_81_4550_2021-01-02 04:34:39.442790       2  11900388   
46092  2_39_81_4550_2021-01-02 15:07:37.007402       2  11900388   
...                                        ...     ...       ...   
51579  2_39_81_4550_2020-12-30 06:35:38.926587       2  11900388   
51580  2_39_81_4550_2020-12-30 18:17:30.573937       2  11900388   
51582  2_39_81_4550_2020-12-30 22:09:41.373141       2  11900388   
51584  2_39_81_4550_2020-12-31 16:17:42.790605       2  11900388   
51586  2_39_81_4550_2020-12-31 17:42:03.239398       2  11900388   

          stationID             timezone   userID  \
46086  2-39-81-4550  America/Los_Angeles   9561.0   
46089  2-39-81-4550  America/Los_Angeles   9283.0   
46090  2-39-81-4550  America/Los_Angeles   9283.0   
46091  2-39-81-4550  America/Los_Angeles    915.0   
46092  2-39-81-4550  America/Los_Angeles   7132.0   
...             ...                  ...      ...   
51579  2-39-81-4550  America/Los_Angeles   1424.0   
51580  2-39-81-4550  America/Los_Angeles    268.0   
51582  2-39-81-4550  America/Los_Angeles  10090.0   
51584  2-39-81-4550  America/Los_Angeles   1039.0   
51586  2-39-81-4550  America/Los_Angeles   5855.0   

                                              userInputs  
46086  [{'WhPerMile': 299, 'kWhRequested': 38.87, 'mi...  
46089  [{'WhPerMile': 319, 'kWhRequested': 28.71, 'mi...  
46090  [{'WhPerMile': 319, 'kWhRequested': 28.71, 'mi...  
46091  [{'WhPerMile': 286, 'kWhRequested': 71.5, 'mil...  
46092  [{'WhPerMile': 274, 'kWhRequested': 27.4, 'mil...  
...                                                  ...  
51579  [{'WhPerMile': 283, 'kWhRequested': 59.43, 'mi...  
51580  [{'WhPerMile': 283, 'kWhRequested': 28.3, 'mil...  
51582  [{'WhPerMile': 290, 'kWhRequested': 14.5, 'mil...  
51584  [{'WhPerMile': 283, 'kWhRequested': 28.3, 'mil...  
51586  [{'WhPerMile': 286, 'kWhRequested': 62.92, 'mi...  

[1125 rows x 12 columns]

We see that this station is always missing the **doneChargingTime** value.

Now we will take care of these values. We see that we cannot derive this value from other values. Since we have **4087** affected rows out of **65037**, thus **6,3%**, dropping all of those would not result in a significant loss.

In [123]:
cleaned_df = formatted_df[formatted_df['doneChargingTime'].isnull()]
cleaned_df

id            connectionTime  \
3756   5ea3848df9af8b428bf97225 2020-04-08 16:36:53+00:00   
3757   5ea4d60df9af8b46573f1ee7 2020-04-09 13:23:52+00:00   
3758   5ea4d60df9af8b46573f1ee8 2020-04-09 13:40:03+00:00   
3759   5ea4d60df9af8b46573f1ee9 2020-04-09 14:01:53+00:00   
3760   5ea4d60df9af8b46573f1eea 2020-04-09 15:01:06+00:00   
...                         ...                       ...   
60411  5cb3d2a8f9af8b4551261e75 2019-03-30 00:32:10+00:00   
61097  5cc8e891f9af8b45d975b819 2019-04-14 16:19:12+00:00   
61103  5cca3a54f9af8b49aaa4cbac 2019-04-15 13:00:59+00:00   
61117  5cca3a54f9af8b49aaa4cbba 2019-04-15 13:52:45+00:00   
65530  5d3f8f96f9af8b7f97666a55 2019-07-13 20:13:18+00:00   

                 disconnectTime doneChargingTime  kWhDelivered  \
3756  2020-04-09 03:43:29+00:00              NaT        25.183   
3757  2020-04-09 17:59:02+00:00              NaT        14.136   
3758  2020-04-09 21:43:22+00:00              NaT        14.203   
3759  2020-04-09 16:33:40+00:00              NaT        13.041   
3760  2020-04-09 23:38:08+00:00              NaT         5.797   
...                         ...              ...           ...   
60411 2019-03-30 02:23:14+00:00              NaT        11.148   
61097 2019-04-14 20:38:39+00:00              NaT        11.678   
61103 2019-04-15 13:55:25+00:00              NaT         3.343   
61117 2019-04-15 14:47:01+00:00              NaT         4.558   
65530 2019-07-13 20:37:40+00:00              NaT         2.218   

                                    sessionID  siteID  spaceID    stationID  \
3756   1_1_194_822_2020-04-08 16:36:53.316264       1  AG-1F12  1-1-194-822   
3757   1_1_193_827_2020-04-09 13:23:51.689520       1  AG-1F02  1-1-193-827   
3758   1_1_178_823_2020-04-09 13:39:55.115497       1  AG-1F08  1-1-178-823   
3759   1_1_178_828_2020-04-09 14:01:53.467693       1  AG-1F10  1-1-178-828   
3760   1_1_193_820_2020-04-09 15:01:06.343454       1  AG-1F04  1-1-193-820   
...                                       ...     ...      ...          ...   
60411  1_1_193_829_2019-03-30 00:32:10.128651       1  AG-1F03  1-1-193-829   
61097  1_1_193_816_2019-04-14 16:19:11.864262       1  AG-1F05  1-1-193-816   
61103  1_1_193_819_2019-04-15 13:00:58.722284       1  AG-1F06  1-1-193-819   
61117  1_1_179_800_2019-04-15 13:52:44.693153       1  AG-3F32  1-1-179-800   
65530  1_1_193_825_2019-07-13 20:13:18.490314       1  AG-1F01  1-1-193-825   

                  timezone  userID  \
3756   America/Los_Angeles   507.0   
3757   America/Los_Angeles   419.0   
3758   America/Los_Angeles  5065.0   
3759   America/Los_Angeles   651.0   
3760   America/Los_Angeles   466.0   
...                    ...     ...   
60411  America/Los_Angeles   194.0   
61097  America/Los_Angeles   364.0   
61103  America/Los_Angeles   651.0   
61117  America/Los_Angeles   194.0   
65530  America/Los_Angeles    76.0   

                                              userInputs  
3756   [{'WhPerMile': 250, 'kWhRequested': 75.0, 'mil...  
3757   [{'WhPerMile': 400, 'kWhRequested': 40.0, 'mil...  
3758   [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
3759   [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
3760   [{'WhPerMile': 300, 'kWhRequested': 6.0, 'mile...  
...                                                  ...  
60411  [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
61097  [{'WhPerMile': 400, 'kWhRequested': 56.0, 'mil...  
61103  [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
61117  [{'WhPerMile': 400, 'kWhRequested': 12.0, 'mil...  
65530  [{'WhPerMile': 301, 'kWhRequested': 9.03, 'mil...  

[4088 rows x 12 columns]